design/ceph/ceph-stretch-cluster.md

```diff
@@ -0,0 +1,224 @@
+
+## Architecture
+
+To enable the stretch cluster based on the Ceph architecture:
+- Rook requires three zones
+- Two zones (A and B) will each run all types of Rook pods. We call these the "data" zones.
+    - Two mons run in each data zone. The OSDs can only connect to the mon in their own zone
+      so we need more than one mon in the data zones.
+- The third zone (Arbiter):
+    - Runs a single mon, called the "tiebreaker" or "arbiter"
+    - No other Rook or Ceph daemons will be run in the arbiter zone
+
+![Stretch Cluster](stretchcluster.png)
+
+A zone is considered down by Ceph when the mons in the zone are all unavailable.
+If the data zones only had a single mon, they would be considered down if the mon
+went down for any reason such as an upgrade. This is another reason two mons are
+required in each of the data zones.
+
+The arbiter zone will commonly contain just a single node that is also a K8s master node,
+although the arbiter zone may certainly contain more nodes.
+
+The type of failure domain used for stretch clusters is always "zone".
+
+Question: Is support is needed for other failure domains such as "datacenter"?
+
+### Latency
+
+Distributed systems are impacted by the network latency between critical components.
+In a stretch cluster, the critical latency is in the Etcd servers configured with Kubernetes.
+K8s only supports latency of up to 5ms (10ms round trip), which is lower than the latency
+requirements for any Ceph components. Ceph mons can handle higher latency, designed for up to
+700ms round trip.
+
+## Design
+
+### Domain Failure Configuration
+
+The topology of the K8s cluster is to be determined by the admin, outside the scope of Rook.
+Rook will simply detect the topology labels that have been added to the nodes.
+
+If the desired failure domain is a "zone", the `topology.kubernetes.io/zone` label should
+be added to the nodes. Any of the [topology labels](https://rook.io/docs/rook/master/ceph-cluster-crd.html#osd-topology)
+supported by OSDs can be used.
+
+In the minimum configuration, two nodes in each data zone would be labeled, while one
+node in the arbiter zone is required.
+
+For example:
+
+```yaml
+topology.kubernetes.io/zone=a
+topology.kubernetes.io/zone=a
+topology.kubernetes.io/zone=b
+topology.kubernetes.io/zone=b
+topology.kubernetes.io/zone=arbiter
+```
+
+### Rook Cluster Design
+
+The core changes to rook are to associate the mons with the required zones.
+
+- Five mons are required. No other count of mons is supported for stretch clusters.
+- One mon will be assigned to the arbiter zone
+- Two mons will each be assigned to the data zones
+  - The two mons within each zone will have node antiaffinity
+
+The new configuration is found under the `stretchCluster` configuration where the
+three zones must be listed and the arbiter zone is identified.
+
+```yaml
+  mon:
+    count: 5
+    allowMultiplePerNode: false
+    stretchCluster:
+      zones:
```

Comment:
So how do we determine whether stretchCluster is enabled or not? Just by looking at the zone length?

Reply:
Yes, I'm thinking if the stretch cluster zones are included, then it's a stretch cluster. If no zones specified it's not a stretch cluster. If anything other than 3 zones, it would be an error.